src/t_zset.c

```diff
@@ -115,15 +115,40 @@ static inline void zslSetNodeHeight(zskiplistNode *x, int height) {
 }
 
 /* Create a skiplist node with the specified number of levels.
- * The SDS string 'ele' is referenced by the node after the call. */
-static zskiplistNode *zslCreateNode(int height, double score, sds ele) {
-    zskiplistNode *zn = zmalloc(sizeof(*zn) + height * sizeof(struct zskiplistLevel));
+ * By embedding elements and levels into the skiplist nodes,
+ * we achieve better cache-friendliness and a more compact memory structure.
+ *
+ * The memory layout is as follows:
+ *
+ *     +-------+------------------+---------+-----+---------+-------------+
+ *     | score | backward-pointer | level-0 | ... | level-N | element-sds |
+ *     +-------+------------------+---------+-----+---------+-------------+
+ */
+static zskiplistNode *zslCreateNode(int height, double score, const_sds ele) {
+    size_t ele_sds_len = ele ? sdslen(ele) : 0;
+    char ele_sds_type = ele ? sdsReqType(ele_sds_len) : 0;
+    size_t ele_sds_size = ele ? sdsReqSize(ele_sds_len, ele_sds_type) : 0;
+    /* Allocate enough space for the node, levels, and the element SDS string.
+     * The +1 is for the header size of the SDS string. */
```

Comment:
This extra byte deserves some more explanation. We've used it in other places where we embedded sds but the reader of this code might not be aware of that. Something like this?

```suggestion
    /* Allocate enough space for the node, levels, and the element SDS string.
     * Before the embedded sds, we include one extra byte representing the sds
     * header size, which is the offset into the embedded sds data where the
     * string content starts. */
```